sonic_y_cable/y_cable.py

```diff
@@ -1097,11 +1129,104 @@ def check_if_nic_lanes_active(physical_port):
 
     if platform_chassis is not None:
         res = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
-        y_cable_validate_read_data(res, 1, physical_port, "nic lanes active")
+        if y_cable_validate_read_data(res, 1, physical_port, "nic lanes active") == EEPROM_READ_DATA_INVALID:
+            return EEPROM_ERROR
         result = res[0]
 
     else:
         helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
         return -1
 
     return result
+
+
+def get_firmware_version(physical_port, target):
+
+    data = bytearray(FIRMWARE_INFO_PAYLOAD_SIZE)
+
+    if platform_chassis is not None:
+        for byteIdx in range(0, FIRMWARE_INFO_PAYLOAD_SIZE):
+            curr_offset = 0xfc * 128 + 128 + byteIdx
+            readOut = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+            if y_cable_validate_read_data(readOut, 1, physical_port, "firmware info") == EEPROM_READ_DATA_INVALID:
+                return EEPROM_ERROR
+            data[byteIdx] = readOut[0]
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
+        return -1
+
+    result = []
+    NUM_MCU_SIDE = 3
+    Addr = target * (FIRMWARE_INFO_PAYLOAD_SIZE / NUM_MCU_SIDE)
+    baseAddr = int(Addr)
+    rev_major_slot1 = struct.unpack_from('<B', data[0 + baseAddr:  1 + baseAddr])[0]
+    rev_minor_slot1 = struct.unpack_from('<B', data[2 + baseAddr:  3 + baseAddr])[0]
+    rev_build_lsb_slot1 = struct.unpack_from('<B', data[4 + baseAddr:  5 + baseAddr])[0]
+    rev_build_msb_slot1 = struct.unpack_from('<B', data[5 + baseAddr:  6 + baseAddr])[0]
+    rev_major_slot2 = struct.unpack_from('<B', data[1 + baseAddr:  2 + baseAddr])[0]
+    rev_minor_slot2 = struct.unpack_from('<B', data[3 + baseAddr:  4 + baseAddr])[0]
+    rev_build_lsb_slot2 = struct.unpack_from('<B', data[6 + baseAddr:  7 + baseAddr])[0]
+    rev_build_msb_slot2 = struct.unpack_from('<B', data[7 + baseAddr:  8 + baseAddr])[0]
+    slotStatus = struct.unpack_from('<B', data[8 + baseAddr:  9 + baseAddr])[0]
```

Comment:
Suggest renaming `slot_status`.

Also, what is this value? Does it represent the active slot?

Reply:
actually slot_status would be used by api caller to determine which slot has run, commit and empty. Something like this. 
`if slot_status & 0x01: runSlot1    = '[*]'
        else:                 runSlot1    = '[ ]'
        if slot_status & 0x02: commitSlot1 = '[*]'
        else:                 commitSlot1 = '[ ]'
        if slot_status & 0x04: emptySlot1  = '[*]'
        else:                 emptySlot1  = '[ ]'
        if slot_status & 0x10: runSlot2    = '[*]'
        else:                 runSlot2    = '[ ]'
        if slot_status & 0x20: commitSlot2 = '[*]'
        else:                 commitSlot2 = '[ ]'
        if slot_status & 0x40: emptySlot2  = '[*]'
        else:                 emptySlot2  = '[ ]'`